### PR TITLE
Support PyTorch

### PR DESCRIPTION
**Pull request by oliverhu**
Support PyTorch

### CODE REVIEW

File: tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java [hunk 1]
```diff
@@ -72,6 +73,8 @@ protected TaskExecutor() throws IOException {
     this.tbPort = this.tbSocket.getLocalPort();
     this.gatewayServerSocket = new ServerSocket(0);
     this.gatewayServerPort = this.gatewayServerSocket.getLocalPort();
+    framework = TonyConfigurationKeys.MLFramework.valueOf(tonyConf.get(TonyConfigurationKeys.FRAMEWORK_NAME,
```
Comment (erwa) [inline tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java]: Can we be consistent and either use `this.` everywhere or not use `this.` everywhere? My personal preference is to omit `this.` and name member variables with a `_` prefix.

File: tony-core/src/main/java/com/linkedin/tony/Constants.java [hunk 2]
```diff
@@ -15,6 +15,12 @@
   public static final String TASK_INDEX = "TASK_INDEX";
   public static final String TASK_NUM = "TASK_NUM";
   public static final String CLUSTER_SPEC = "CLUSTER_SPEC";
+
+  // PyTorch constants
```
Comment (erwa) [inline tony-core/src/main/java/com/linkedin/tony/Constants.java]: I think you should also add a comment on line 13 saying those are TensorFlow constants

File: tony-core/src/main/java/com/linkedin/tony/TonyApplicationMaster.java [hunk 3]
```diff
@@ -160,6 +160,9 @@
   // Failure conditions
   private boolean jobFailed;
 
+  // Handle different machine frameworks
+  private TonyConfigurationKeys.MLFramework framework;
```
Comment (erwa) [inline tony-core/src/main/java/com/linkedin/tony/TonyApplicationMaster.java]: Can you import `TonyConfigurationKeys.MLFramework` and just use `MLFramework` here and elsewhere in this file?

**Comment by erwa** [comment]
Looks like several of the CI tests failed due to 
```
Gradle suite > Gradle test > com.linkedin.tony.TestTaskExecutor.testTaskExecutorConf FAILED
    java.lang.IllegalArgumentException: No enum constant com.linkedin.tony.TonyConfigurationKeys.MLFramework.tensorflow
        at java.lang.Enum.valueOf(Enum.java:238)
        at com.linkedin.tony.TonyConfigurationKeys$MLFramework.valueOf(TonyConfigurationKeys.java:8)
        at com.linkedin.tony.TaskExecutor.<init>(TaskExecutor.java:76)
        at com.linkedin.tony.TestTaskExecutor.testTaskExecutorConf(TestTaskExecutor.java:38)
```

Can you fix this?

**Comment by oliverhu** [comment]
@erwa  yah, just pushed a commit to fix that.

File: tony-examples/mnist-pytorch/mnist_distributed.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    parser.add_argument(
+        '--init-method',
+        '-i',
+        type=str,
+        default=os.environ.get('INIT_METHOD', 'tcp://127.0.0.1:23456'),
+        help='URL specifying how to initialize the package.')
+    parser.add_argument(
+        '--rank', '-r',
+        type=int,
+        defailt=int(os.environ.get('RANK')),
```
Comment (erwa) [inline tony-examples/mnist-pytorch/mnist_distributed.py]: defailt -> default (I'm surprised this didn't fail at runtime)

File: tony-examples/mnist-pytorch/mnist_distributed.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        help='Rank of the current process.')
+    parser.add_argument(
+        '--world-size',
+        '-s',
+        type=int,
+        defailt=int(os.environ.get('WORLD')),
```
Comment (erwa) [inline tony-examples/mnist-pytorch/mnist_distributed.py]: same

File: tony-core/src/main/java/com/linkedin/tony/TonyApplicationMaster.java [hunk 6]
```diff
@@ -764,6 +769,13 @@ public String registerWorkerSpec(String taskId, String spec) throws IOException
         LOG.info("Received cluster spec registration request from task " + taskId + " with spec: " + spec);
         task.setHostPort(spec);
         registeredTasks.add(taskId);
+
+        // Use chief worker as coordinator.
+        if (taskId.equals(COORDINATOR_ID) && framework == MLFramework.PYTORCH) {
+          // Hard coded to use tcp:// as backend.
```
Comment (erwa) [inline tony-core/src/main/java/com/linkedin/tony/TonyApplicationMaster.java]: Can we add a TODO to support other backends like Gloo, MPI, and NCCL? Gloo and NCCL will allow us to take advantage of GPU features like direct memory access/write while TCP will not.

File: tony-core/src/main/java/com/linkedin/tony/TonyConfigurationKeys.java [hunk 7]
```diff
@@ -5,6 +5,20 @@
 package com.linkedin.tony;
 
 public class TonyConfigurationKeys {
+  public enum MLFramework {
+    TENSORFLOW("tensorflow"),
+    PYTORCH("pytorch");
+
+    private final String frameworkName;
```
Comment (erwa) [inline tony-core/src/main/java/com/linkedin/tony/TonyConfigurationKeys.java]: I think this String field is unnecessary. By default, if you use
```
  public enum MLFramework {
    TENSORFLOW,
    PYTORCH
  }
```

MLFramework.toString() will return TENSORFLOW or PYTORCH.

**Comment by oliverhu** [comment]
Refactoring out all the tf would be a fairly big change, will do that in a follow up rb.
